megamek/src/megamek/common/net/SocketServer.java

```diff
@@ -0,0 +1,159 @@
+     *        wish to be notified of this event
+     * @Note: This method will be invoked from a different thread and it is the responsibility
+     *        of the subclass to manage any cross-thread synchronization needed.
+     */
+    protected void onPacketReceived(IConnection connection, Packet packet) {
+    }
+
+    private ConnectionListener connectionListener = new ConnectionListener() {
+        @Override
+        public void connected(ConnectedEvent e) {
+        }
+
+        @Override
+        public void disconnected(DisconnectedEvent e) {
+            IConnection conn = e.getConnection();
+            onConnectionClose(conn);
+
+            ConnectionHandler ch = connectionHandlers.get(conn.getId());
+            if (ch != null) {
+                ch.signalStop();
+                connectionHandlers.remove(conn.getId());
+            }
+        }
+
+        @Override
+        public void packetReceived(PacketReceivedEvent e) {
+            onPacketReceived(e.getConnection(), e.getPacket());
+        }
+    };
+
+    private MMLogger getLogger() {
+        if (null == logger) {
+            logger = DefaultMmLogger.getInstance();
+        }
+
+        return logger;
+    }
+
+    private void logInfo(String methodName, String message) {
+        getLogger().log(getClass(), methodName, LogLevel.INFO, message);
+    }
+
+    protected void listen(int port) throws IOException {
+        serverSocket = new ServerSocket(port);
+
+        listenThread = new Thread(this, "Socket Server");
+        listenThread.start();
+        onListen();
+    }
+
+    protected int getLocalPort() {
+        if (serverSocket == null) {
+            return 0;
+        }
+
+        return serverSocket.getLocalPort();
+    }
+
+    protected void stopNetwork() throws Exception {
+        serverSocket.close();
+        onListenEnded();
+    }
+
+    @Override
+    public void run() {
+        final String METHOD_NAME = "run()";
+        logInfo(METHOD_NAME, "s: listening for clients...");
+
+        Thread currentThread = Thread.currentThread();
+        while (listenThread == currentThread) {
+            try {
+                Socket s = serverSocket.accept();
+
+                IConnection connection = ConnectionFactory.getInstance()
+                        .createServerConnection(s, 0);
+
+                // call out to subclass; this should set a valid ID on the connection
+                onConnectionOpen(connection);
+
+                connection.addConnectionListener(connectionListener);
+
+                ConnectionHandler ch = new ConnectionHandler(connection);
+                Thread newConnThread = new Thread(ch, "Connection " + connection.getId());
+                newConnThread.start();
+                connectionHandlers.put(connection.getId(), ch);
+            } catch (InterruptedIOException iioe) {
+                // ignore , just SOTimeout blowing..
+            } catch (IOException ex) {
+
```

Comment:
should these too be ignored or logged?

Reply:
I believe those were in the original code -- I just lifted the existing code out of the `Server` class and packaged it into this one